aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -1180,5 +1182,58 @@ Tensor _reshape_nested(const Tensor& self, IntArrayRef proposed_shape) {
   return output;
 }
 
+Tensor slice_nested(
+    const Tensor& self,
+    int64_t dim,
+    c10::optional<int64_t> start,
+    c10::optional<int64_t> end,
+    int64_t step) {
+  int64_t ndim = self.dim();
+  if (ndim == 0) {
+    TORCH_CHECK_INDEX(false, "slice() cannot be applied to a 0-dim tensor.");
+  }
+  dim = maybe_wrap_dim(dim, ndim);
+
+  // will error if the dimension is jagged
+  auto dim_size = self.size(dim);
+
+  TORCH_CHECK(step > 0, "slice step must be positive");
+
+  auto* nt_impl = get_nested_tensor_impl(self);
+  const int64_t ntensors = nt_impl->size(0);
+  const auto& sizes = nt_impl->get_nested_size_tensor();
+  const auto& strides = nt_impl->get_nested_stride_tensor();
+  const std::vector<int64_t>& offsets = nt_impl->get_offsets();
+
+  // special case for empty NT
+  if (ntensors == 0) {
+    return create_nested_view_tensor(self, sizes.clone(), strides.clone(), std::vector<int64_t>(offsets));
+  }
+
+  int64_t start_val, end_val;
+  std::tie(start_val, end_val) = get_slice_range(start, end, dim_size);
+  auto len = end_val - start_val;
+
+  // special case for slicing the tensor components themselves when dim=0
+  auto new_sizes = (dim > 0) ? sizes.clone() : sizes.slice(0, start_val, end_val).clone();
+  auto new_strides = (dim > 0) ? strides.clone() : strides.slice(0, start_val, end_val).clone();
+  auto new_offsets = (dim > 0) ? std::vector<int64_t>(offsets) :
+      std::vector<int64_t>(offsets.begin() + start_val, offsets.begin() + end_val + 1);
+  if (dim > 0) {
+    // account for the implicit batch dim
+    --dim;
+    for (int64_t i : c10::irange(ntensors)) {
+      int64_t *size_ptr = new_sizes[i].data_ptr<int64_t>();
+      int64_t *stride_ptr = new_strides[i].data_ptr<int64_t>();
+
+      new_offsets[i] = offsets[i] + start_val * stride_ptr[dim];
```

Comment:
@albanD I might forward fix: #84145 and change the offsets size back to n_tensors, but I liked my code changes so want to keep those.. The changes I made here after rebasing should be valid however. The problem is that the invariant does not hold when you slice dim != 0 meaning the n+1 element doesn't maintain any semantic value either.

Reply:
actually just going to update the this PR in the stack

Reply:
Update. Since the last review, I added the `is_nested` path in slice and changed back the num elements in offsets from n_tensors +1 -> n_tensors